### PR TITLE
Sbus output

### PR DESCRIPTION
**Pull request by gongtao0607**
This PR implements #165

It defines 18 (16 full-scale and 2 on-off) -channel S.Bus output. Each channel can be mapped from an arbitrary RX channel, Mixer output, servo output or motor output.

Quick guide:
1. Enable an sbus output serial port:
    `serial 1 262144 115200 57600 0 115200` (on Flydragon, this is the "F.Port". The baud rate doesn't matter.)
2. Configure channels:
```
set sbus_out_source_type  = 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
set sbus_out_source_index = 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17
set sbus_out_source_range_low = 1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000
set sbus_out_source_range_high = 2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000,2000
set sbus_out_frame_rate = 100
set sbus_out_pinswap = OFF
```
(This will mirror all RX channels to SBus out channels)

For each channel:
 * `sbus_out_source_type` can be set to 0 (None),1 (RX), 2 (Mixer), 3 (Servo), or 4 (Motor)
 * `sbus_out_source_index` sets the source channel, e.g., RX channel **14**, mixer rule **4**, servo **2**. Index are all 0-based.
 * `sbus_out_source_range_low ` and `sbus_out_source_range_high ` setup the value mapping. 
     * For RX channels or servos (wideband), use 1000, 2000.
     * For mixer rules, use -1000, 1000.
     * For narrowband servos, use 500, 1000.
     * For motors, use 0, 1000.
     * Or you can customize your own mapping.
 
`sbus_out_frame_rate` is to set the output rate. Common receivers support 50 - 100Hz. You are allowed to set 25-250Hz.

### CODE REVIEW

**Comment by rotorflight** [comment]
RF-4.4.x branch is a release branch, and is generally in feature freeze.
It REALLY MEANS no new features are allowed in it.
This is a new feature.
Please move it to the master branch, where it belongs.

**Comment by robthomson** [comment]
I have reverted the 3 commits in this PR that included all the google test code.

This is up on:

https://github.com/rotorflight/rotorflight-firmware/pull/193

**Comment by robthomson** [comment]
That's better.

Thank you

**Comment by robthomson** [comment]
I have tested this code and all seems pretty stable.   Quite flexible all round.

File: src/main/drivers/sbus_output.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    frame->chan15 = channels[15];
+
+    frame->flags = channels[16] ? 0x01 : 0x00;
+    frame->flags += channels[17] ? 0x02 : 0x00;
+    // Other flags?
+
+    frame->endByte = 0;
+}
+
+static float sbusOutGetRX(uint8_t channel) {
```
Comment (rotorflight) [inline src/main/drivers/sbus_output.c:64]: Please follow the formatting convention on functions.
The opening { should be on the next line.

File: src/main/fc/tasks.c [hunk 2]
```diff
@@ -568,5 +575,10 @@ void tasksInit(void)
     setTaskEnabled(TASK_TELEMETRY_SBUS2, useSBUS2);
 #endif
 
+#ifdef USE_SBUS_OUTPUT
+    rescheduleTask(TASK_SBUS_OUTPUT, TASK_PERIOD_HZ(sbusOutConfig()->sbusRate));
+    setTaskEnabled(TASK_SBUS_OUTPUT, true);
```
Comment (rotorflight) [inline src/main/fc/tasks.c:580]: It appears the SBUS task is always enabled?
This is not good. Please enable it only if the output is configured.

File: src/main/msp/msp.c [hunk 3]
```diff
@@ -1696,6 +1697,18 @@ static bool mspProcessOutCommand(int16_t cmdMSP, sbuf_t *dst)
         break;
 #endif
 
+#ifdef USE_SBUS_OUTPUT
+    case MSP_SBUS_OUTPUT_SETTINGS:
```
Comment (rotorflight) [inline src/main/msp/msp.c:1701]: The convention is to call it _CONFIG, not _SETTINGS

File: src/main/pg/pg_ids.h [hunk 4]
```diff
@@ -85,6 +85,7 @@
 // Driver configuration
 #define PG_DRIVER_PWM_RX_CONFIG 100 // does not exist in betaflight
 #define PG_DRIVER_FLASHCHIP_CONFIG 101 // does not exist in betaflight
+#define PG_DRIVER_SBUS_OUT_CONFIG 102 // does not exist in betaflight
```
Comment (rotorflight) [inline src/main/pg/pg_ids.h:88]: Please use 1005

File: src/main/msp/msp_protocol.h [hunk 5]
```diff
@@ -193,6 +193,8 @@
 #define MSP_SET_GOVERNOR_PROFILE             149
 #define MSP_LED_STRIP_SETTINGS               150
 #define MSP_SET_LED_STRIP_SETTINGS           151
+#define MSP_SBUS_OUTPUT_SETTINGS             152
+#define MSP_SET_SBUS_OUTPUT_SETTINGS         153
```
Comment (rotorflight) [inline src/main/msp/msp_protocol.h:197]: Please rename to _CONFIG

File: src/main/drivers/sbus_output.c [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+void sbusOutInit() {
+    const serialPortConfig_t *portConfig =
+        findSerialPortConfig(FUNCTION_SBUS_OUT);
+
+    if (!portConfig) {
+        sbusOutPort = NULL;
+        return;
+    }
+
+    sbusOutPort = openSerialPort(
```
Comment (rotorflight) [inline src/main/drivers/sbus_output.c:152]: Need to add the pinswap CLI option

File: src/main/cli/settings.c [hunk 7]
```diff
@@ -1738,6 +1739,14 @@ const clivalue_t valueTable[] = {
...
     { "box_user_4_name", VAR_UINT8 | MASTER_VALUE | MODE_STRING, .config.string = { 1, MAX_BOX_USER_NAME_LENGTH, STRING_FLAGS_NONE }, PG_MODE_ACTIVATION_CONFIG, offsetof(modeActivationConfig_t, box_user_4_name) },
 #endif
+
+#ifdef USE_SBUS_OUTPUT
+    { "sbus_out_source_type",  VAR_UINT8 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sourceType) },
+    { "sbus_out_source_index", VAR_UINT8 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sourceIndex) },
+    { "sbus_out_min",          VAR_INT16 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, min) },
+    { "sbus_out_max",          VAR_INT16 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, max) },
+    { "sbus_out_sbus_rate",    VAR_UINT8 | MASTER_VALUE, .config.minmaxUnsigned = {25, 250}, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sbusRate) },
+#endif
```
Comment (rotorflight) [inline src/main/cli/settings.c:1749]: pinswap option missing

File: src/main/pg/sbus_output.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma once
+
+#include "common/utils.h"
+#include "pg/pg.h"
+
+#define SBUS_OUT_CHANNELS 18
+
+typedef enum {
+    SBUS_OUT_SOURCE_RX = 0,
```
Comment (rotorflight) [inline src/main/pg/sbus_output.h:26]: Please add SBUS_OUT_SOURCE_NONE = 0, for nothing on the channel

File: src/main/cli/settings.c [hunk 9]
```diff
@@ -1738,6 +1739,14 @@ const clivalue_t valueTable[] = {
     { "box_user_3_name", VAR_UINT8 | MASTER_VALUE | MODE_STRING, .config.string = { 1, MAX_BOX_USER_NAME_LENGTH, STRING_FLAGS_NONE }, PG_MODE_ACTIVATION_CONFIG, offsetof(modeActivationConfig_t, box_user_3_name) },
     { "box_user_4_name", VAR_UINT8 | MASTER_VALUE | MODE_STRING, .config.string = { 1, MAX_BOX_USER_NAME_LENGTH, STRING_FLAGS_NONE }, PG_MODE_ACTIVATION_CONFIG, offsetof(modeActivationConfig_t, box_user_4_name) },
 #endif
+
+#ifdef USE_SBUS_OUTPUT
+    { "sbus_out_source_type",  VAR_UINT8 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sourceType) },
+    { "sbus_out_source_index", VAR_UINT8 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sourceIndex) },
+    { "sbus_out_min",          VAR_INT16 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, min) },
```
Comment (rotorflight) [inline src/main/cli/settings.c:1746]: sbus_out_source_range_low

File: src/main/cli/settings.c [hunk 10]
```diff
@@ -1738,6 +1739,14 @@ const clivalue_t valueTable[] = {
...
+    { "sbus_out_max",          VAR_INT16 | MASTER_VALUE | MODE_ARRAY, .config.array.length = SBUS_OUT_CHANNELS, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, max) },
```
Comment (rotorflight) [inline src/main/cli/settings.c:1747]: sbus_out_source_range_high

File: src/main/cli/settings.c [hunk 11]
```diff
@@ -1738,6 +1739,14 @@ const clivalue_t valueTable[] = {
...
+    { "sbus_out_sbus_rate",    VAR_UINT8 | MASTER_VALUE, .config.minmaxUnsigned = {25, 250}, PG_DRIVER_SBUS_OUT_CONFIG, offsetof(sbusOutConfig_t, sbusRate) },
```
Comment (rotorflight) [inline src/main/cli/settings.c:1748]: sbus_out_frame_rate

File: src/main/drivers/sbus_output.c [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    if (channel < MAX_SUPPORTED_SERVOS)
+        return getServoOutput(channel);
+    return 0;
+}
+
+// Note: this returns 0 - 1000 range (or -1000 - 1000 for bidirectional
+// motor)
+static float sbusOutGetMotor(uint8_t channel) {
+    if (channel < MAX_SUPPORTED_MOTORS)
+        return getMotorOutput(channel);
```
Comment (rotorflight) [inline src/main/drivers/sbus_output.c:87]: I think getMotorOutput returns 0..1 or -1..1. So you should *1000.0f. Please check.

File: src/main/drivers/sbus_output.c (hunk 1 above)
Comment (gongtao0607) [reply src/main/drivers/sbus_output.c:64]: Done

File: src/main/drivers/sbus_output.c (hunk 12 above)
Comment (gongtao0607) [reply src/main/drivers/sbus_output.c:92]: https://github.com/rotorflight/rotorflight-firmware/blob/f722bf581e238850e402076d1ec337821c5867b7/src/main/flight/motors.c#L94
This function returns 1000x 0..1 or -1..-1

File: src/main/cli/settings.c (hunk 11 above)
Comment (gongtao0607) [reply src/main/cli/settings.c]: changed

File: src/main/cli/settings.c (hunk 10 above)
Comment (gongtao0607) [reply src/main/cli/settings.c]: Changed

File: src/main/cli/settings.c (hunk 9 above)
Comment (gongtao0607) [reply src/main/cli/settings.c]: changed

File: src/main/pg/sbus_output.h (hunk 8 above)
Comment (gongtao0607) [reply src/main/pg/sbus_output.h]: Added

File: src/main/cli/settings.c (hunk 7 above)
Comment (gongtao0607) [reply src/main/cli/settings.c:1750]: Added

File: src/main/drivers/sbus_output.c (hunk 6 above)
Comment (gongtao0607) [reply src/main/drivers/sbus_output.c:168]: Added sbusOutConfig_t->pinSwap

File: src/main/msp/msp_protocol.h (hunk 5 above)
Comment (gongtao0607) [reply src/main/msp/msp_protocol.h]: Done

File: src/main/pg/pg_ids.h (hunk 4 above)
Comment (gongtao0607) [reply src/main/pg/pg_ids.h]: Done

File: src/main/msp/msp.c (hunk 3 above)
Comment (gongtao0607) [reply src/main/msp/msp.c]: Done

File: src/main/fc/tasks.c (hunk 2 above)
Comment (gongtao0607) [reply src/main/fc/tasks.c]: Added sbusOutIsEnabled() to check

**Comment by gongtao0607** [comment]
Please take a look at [b4c3f39](https://github.com/rotorflight/rotorflight-firmware/pull/190/commits/b4c3f39760ea80b1d07c909be68ae5e08faccd55) for the update

File: src/main/fc/tasks.c [hunk 13]
```diff
@@ -568,5 +575,12 @@ void tasksInit(void)
     setTaskEnabled(TASK_TELEMETRY_SBUS2, useSBUS2);
 #endif
 
+#ifdef USE_SBUS_OUTPUT
+    rescheduleTask(TASK_SBUS_OUTPUT, TASK_PERIOD_HZ(sbusOutConfig()->frameRate));
+    if (sbusOutIsEnabled()) {
```
Comment (rotorflight) [inline src/main/fc/tasks.c:580]: Following the convention....
```
setTaskEnabled(TASK_SBUS_OUTPUT, sbusOutIsEnabled());
```

File: src/main/drivers/sbus_output.c [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    frame->chan9 = channels[9];
+    frame->chan10 = channels[10];
+    frame->chan11 = channels[11];
+    frame->chan12 = channels[12];
+    frame->chan13 = channels[13];
+    frame->chan14 = channels[14];
+    frame->chan15 = channels[15];
+
+    frame->flags = channels[16] ? 0x01 : 0x00;
+    frame->flags += channels[17] ? 0x02 : 0x00;
```
Comment (rotorflight) [inline src/main/drivers/sbus_output.c:59]: When dealing with bitmasks, don't use +, but |
```
frame->flags = channels[16] ? BIT(0) : 0;
frame->flags |= channels[17] ? BIT(1) : 0;
```

**Review by rotorflight — CHANGES_REQUESTED** [review]
Sorry about being pedantic...

File: src/main/fc/tasks.c (hunk 13 above)
Comment (gongtao0607) [reply src/main/fc/tasks.c]: Fixed

File: src/main/drivers/sbus_output.c (hunk 14 above)
Comment (gongtao0607) [reply src/main/drivers/sbus_output.c]: Fixed

**Comment by gongtao0607** [comment]
Please take another look at [adcf08a](https://github.com/rotorflight/rotorflight-firmware/pull/190/commits/adcf08add3d6041377dee4bdb311cef9153a071d) Thanks!

File: src/main/drivers/sbus_output.c (hunk 14 above)
Comment (rotorflight) [reply src/main/drivers/sbus_output.c]: Please use the BIT() marcro too!

Comment (gongtao0607) [reply src/main/drivers/sbus_output.c]: Missed it. [5cfeaea](https://github.com/rotorflight/rotorflight-firmware/pull/190/commits/5cfeaeaaf89d81549a126bd9bbd13597f22ca67a)

File: src/main/drivers/sbus_output.c [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+
+    if (!portConfig) {
+        sbusOutPort = NULL;
+        return;
+    }
+
+    sbusOutPort = openSerialPort(
+        portConfig->identifier, FUNCTION_SBUS_OUT, NULL, NULL, 100000, MODE_TX,
+        SERIAL_STOPBITS_2 | SERIAL_PARITY_EVEN | SERIAL_INVERTED |
+                SERIAL_BIDIR | sbusOutConfig()->pinSwap
```
Comment (rotorflight) [inline src/main/drivers/sbus_output.c:171]: I just noticed this.
SERIAL_BIDIR with MODE_TX does not make much sense.
Can you explain your thinking?

**Comment by gongtao0607** [comment]
MSP changes excluded

File: src/main/drivers/sbus_output.c (hunk 15 above)
Comment (gongtao0607) [reply src/main/drivers/sbus_output.c]: No idea what I was thinking. [b71de22](https://github.com/rotorflight/rotorflight-firmware/pull/190/commits/b71de22666552909d52485ad9dcffc485d85498f)

**Review by rotorflight — APPROVED** [review]

**Pull request merged**